mcnode/node.go

```diff
@@ -323,3 +343,106 @@ func (node *Node) doShutdown() {
 	node.ds.Close()
 	os.Exit(0)
 }
+
+func (auth *PeerAuth) fromJSON(rmap map[string]interface{}) error {
+	auth.mx.Lock()
+	defer auth.mx.Unlock()
```

Comment:
It's more like `unwind-protect` at function-scope.
Each function has an unwind stack, where defers pushes thunks. And at the function epilogue, the unwind stack is poped and the thunks are executed.